src/internal/pjsdb/job_crud.go

```diff
@@ -171,19 +171,34 @@ func CancelJob(ctx context.Context, tx *pachsql.Tx, id JobID) ([]JobID, error) {
 	return ids, nil
 }
 
+// WalkAlgorithm is an enumerator for walking algorithms.
+type WalkAlgorithm int32
+
+const (
+	LevelOrder        = WalkAlgorithm(0)
+	PreOrder          = WalkAlgorithm(iota)
+	MirroredPostOrder = WalkAlgorithm(iota)
+)
+
 // WalkJob walks from job 'id' down to all of its children.
-func WalkJob(ctx context.Context, tx *pachsql.Tx, id JobID) ([]Job, error) {
+func WalkJob(ctx context.Context, tx *pachsql.Tx, id JobID, algo WalkAlgorithm) ([]Job, error) {
 	pctx.Child(ctx, "walkJob")
 	job, err := GetJob(ctx, tx, id)
 	if err != nil {
 		return nil, errors.Wrap(err, "walk job")
 	}
-	records := make([]jobRecord, 0)
-	if err = sqlx.SelectContext(ctx, tx, &records, recursiveTraverseChildren+selectJobRecordPrefix+`
-		INNER JOIN children c ON j.id = c.id
-		GROUP BY j.id
-		ORDER BY MIN(depth) ASC;
-	`, job.ID); err != nil {
+	var records []jobRecord
+	var walker func(ctx context.Context, tx *pachsql.Tx, id JobID) ([]jobRecord, error)
+	switch algo {
```

Comment:
a comment `//exhaustive:enforce` will make the linter fail if you add a new WalkAlgorithm and don't update this switch statement.